### PR TITLE
Fixed invalid shapes and lines in GeoWidget, fix #1124

### LINKED ISSUE #1124: Invalid geoshape
Last issue of the week after a great course in Montpelier with 12 people working on Biodiversity and landscape conservation :-)

We face a bug using geoshape.

When people are saving the geoshape after 2 points (they see a line on the widget), the object created is not a line (that's normal with geoshape) but the created polygon is a invalid polygon, with 3 nodes (node 1 and node 3 are the same one) and no interior.

Actually we use a workaround in PostGIS : if the line contains 3 node and if it is closed, we create a line instead of a polygon.

I reported another issue about storing geo data into WKT format and about the size of the column created by aggregate.
- https://github.com/opendatakit/collect/issues/1123
- https://github.com/opendatakit/opendatakit/issues/1265#issuecomment-307324702

### PR DESCRIPTION
**Pull request by huangyz0918**
Closes #1124 

#### What has been done to verify that this works as intended?
I have tested in Android 4.2 4.4 5.1 6.0 7.1 8.1, and it worked well.
#### Why is this the best possible solution? Were any other approaches considered?
~I found every geometry has duplicated nodes, whatever how many points it has.
This is my tests:~
```java
//Two points, note1 and note3 are duplicated
37.04014527453826 89.59821428571429 0.0 0.0;//note1
42.66397254960217 120.83705357142856 0.0 0.0;//note2
37.04014527453826 89.59821428571429 0.0 0.0;//note3

//Three points, note1 and note4 are duplicated
34.390074413373064 93.515625 0.0 0.0;//note1
39.29058236772298 110.08928571428571 0.0 0.0;//note2
30.188548455057692 107.27678571428571 0.0 0.0;//note3
34.390074413373064 93.515625 0.0 0.0;//note4

// The same with other cases:
38.90080485615609 90.60267857142856 0.0 0.0;
37.95657349860099 112.70089285714285 0.0 0.0;
34.803494210753584 116.96986607142856 0.0 0.0;
29.622585468269065 112.60044642857144 0.0 0.0;
38.90080485615609 90.60267857142856 0.0 0.0;
..... .....
```
~We cannot collect 3 points when user actually selected two points. (an example). Valid geoshape doesn't have duplicated nodes so I removed them and make it looks good. But the map SDK needs a duplicated node in order to draw the shape correctly, so I added the extra node outside the data-saving method when reviewing the geoshape.~

~I think it's the best approach to address this issue:~

| Shape | Map View  | Message |
| :------------: |:---------------:|:-----:|
| Invalid  shape (line or a point)  |  |  |
| Valid shape (>= 3 nodes)  |  |  |

#### Are there any risks to merging this code? If so, what are they?
No.
#### Do we need any specific form for testing your changes? If so, please attach one.
No.
#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
N/A

### CODE REVIEW

**Comment by huangyz0918** [comment]
@mathieubossaert please review

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 1]
```diff
@@ -754,6 +757,16 @@ protected void onActivityResult(int requestCode, int resultCode, final Intent in
                 break;
             case RequestCodes.GEOSHAPE_CAPTURE:
                 String gshr = intent.getStringExtra(GEOSHAPE_RESULTS);
+                String[] gshrs = gshr.split(";");
+                // make sure the shape has at least three notes
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: “notes” is meant to be nodes or something?

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 2]
```diff
@@ -754,6 +757,16 @@ protected void onActivityResult(int requestCode, int resultCode, final Intent in
...
+                if (gshrs.length < 4) {
+                    ToastUtils.showShortToastInMiddle("You have to set at least 3 points !");
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: English doesn’t put a space before “!”. Is one even needed here?

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java (hunk 1 above)
Comment (huangyz0918) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: Thank you for your careful inspection! I'm not a native english speaker so there will be some spelling mistakes sometimes. Will fix these soon !

**Comment by huangyz0918** [comment]
| Forms | GeoShap | GeoTrace |
| :------------: |:---------------:|:-----:|
| Invalid  shape  |  |  |
| Invalid  line   | N/A |  |

please review @lognaturel

**Comment by lognaturel** [comment]
Looking good! One thing I noticed is that if I only enter 2 points for geoshape I get the expected toast but a repeat of the first point is still added! That means I can then tap save again and it will succeed. Or if I do add another point it's not attached to the second point but to the third one which matches the first. When I enter two points, I should still have two points after the toast is displayed.

**Comment by huangyz0918** [comment]
@lognaturel Fixed.

**Comment by lognaturel** [comment]
@mmarciniak90 If you have a chance for some extra testing, it would be great to get this great fix by @huangyz0918 in the next release.

**Comment by mmarciniak90** [comment]
@huangyz0918 @lognaturel I noticed crash just when I click on a map to add the first point on Android 4.1, 4.2, 4.4

**Comment by huangyz0918** [comment]
@mmarciniak90 I tested it in `master`, and this issue also exists.
The error message is:

```java
    Process: org.odk.collect.android, PID: 1992
    android.content.res.Resources$NotFoundException: File res/drawable/ic_place.xml from drawable resource ID #0x7f0800a6. If the resource you are trying to use is a vector resource, you may be referencing it in an unsupported way. See AppCompatDelegate.setCompatVectorFromResourcesEnabled() for more info.
        at android.content.res.Resources.loadDrawable(Resources.java:2096)
        at android.content.res.Resources.getDrawable(Resources.java:700)
        at android.support.v4.content.ContextCompat.getDrawable(ContextCompat.java:360)

```
Thank your for your tests, and I think we can open another issue for that.

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
Good catch, @mmarciniak90 and @huangyz0918. Looks like we're running into the same issue as in #2204 -- it seems we may want to revert the change to vectors and wait on it until we stop supporting Android 4.x.

@huangyz0918 I'll let you file the issue if you want or if not I'll do it tomorrow! 🙏

**Comment by grzesiek2010** [comment]
It wasn't really a problem with vectors but with used context https://github.com/opendatakit/collect/pull/2239

**Comment by lognaturel** [comment]
Ooh, awesome, @grzesiek2010. I should have dug a little deeper, I didn't realize they were added programmatically. Thanks for the fix!!

**Pull request merged**